### DISCUSSION TITLE
DropDownAction dynamische Auswahlliste

### DISCUSSION THREAD

**Discussion by Bernd9000** [Zefau/ioBroker.jarvis#2632] [General]
Bisher habe ich für ein DropDownAction das Feld "Anzeigevariante" mit den Werten gefüllt die Auswählbar sein sollen.

Problem:
Was mache ich bei einer dynamischen Liste ?
Kann ich auch andere Quellen angeben ? z.B. die `states` Werte von einem IOB Datenpunkt ?

Konkret benutze ich den Home Connect Adapter um meine Waschmaschine zu steuern.
Hier sind z.B. die Waschprogramme/Optionen dynamisch angelegt.

**Comment by mcuiobroker** [comment]
Wie oft ändert sich denn die Liste der Programme?

**Comment by Bernd9000** [comment]
Sie ändert sich fast jedes mal, je nachdem welches Waschprogramm ich nehme.
Es geht dabei z.B. um Temperaturwahl und Schleuderdrehzahl.
Bei Wolle z.B. darf ich von kalt bis 90°C alles nehmen, bei Seide nur Kalt bis 30°C
Es sind also nicht alle Werte erlaubt, Sie werden dynamisch durch das gewählte Waschprogramm geändert.

So wie das DropDown Feld im IOB unter Objects müßte das DropDownAction sein.
Hier erscheinen nur die Werte, die in den states abgelegt wurden.

![1](https://github.com/Zefau/ioBroker.jarvis/assets/77005685/8a561c8d-cf74-4108-9ba5-df919b736b03)


Objectdaten:

{
  "type": "state",
  "common": {
    "name": "LaundryCare_Washer_Option_SpinSpeed",
    "type": "mixed",
    "role": "indicator",
    "write": true,
    "read": true,
    "unit": "",
    "states": {
      "LaundryCare.Washer.EnumType.SpinSpeed.RPM800": "RPM800",
      "LaundryCare.Washer.EnumType.SpinSpeed.RPM1400": "RPM1400",
      "LaundryCare.Washer.EnumType.SpinSpeed.RPM1000": "RPM1000",
      "LaundryCare.Washer.EnumType.SpinSpeed.RPM1200": "RPM1200"
    }
  },
  "native": {},
  "from": "system.adapter.homeconnect.0",
  "user": "system.user.admin",
  "ts": 1710773862264,
  "_id": "homeconnect.0.484010534615005556.programs.selected.options.LaundryCare_Washer_Option_SpinSpeed",
  "acl": {
    "object": 1636,
    "state": 1636,
    "owner": "system.user.admin",
    "ownerGroup": "system.group.administrator"
  }
}

**Comment by mcuiobroker** [comment]
Aber die auswählbaren Programme ändern sich ja nicht, oder?
Man kann an der Maschine immer nur die Schleuder-Programme RPM800, usw auswählen, nur nicht zum ausgewählten Wasch-Programm.
In sofern würde sich auch nicht die `states` ändern?
Oder werden in den zugehörigen DPs immer nur die States angezeigt, die im ausgewählten Programm genutzt werden können?

**Comment by Bernd9000** [comment]
Das letzte stimmt, es gibt zum Schluß nur ein Datenpunkt für z.B. die Schleuderdrehzahl und in diesem Datenpunkt werden die "states" verändert, so das man nur bestimmte Drehzahlen auswählen kann.

**Comment by mcuiobroker** [comment]
Wie viele Geräte-DPs sind es ~ und sind diese alle in einem Gerät zusammengefasst, oder auf unterschiedliche Geräte verteilt?

Passt du die Anzeigevariante dann nochmal an, evtl da dir RPM800 nicht gefällt?

Ich denke mal laut:
- es wäre ein zusätzliches javascript-Programm
- Parameter müsste man setzen -> DPs? oder Waschmaschinen-DP_Ordner?
- bei Änderung von states innerhalb der DPs -> Änderung in jarvis devices setzen

**Comment by Bernd9000** [comment]
Gut, also nehme ich mal an das in der "Anzeigevariante" in jarvis kein "Binding" möglich ist?
Hast Du ein Beispiel wie ich das im jarvis devices mit einem script ändern kann ?

**Comment by mcuiobroker** [comment]
Binding -> Müsste man probieren, weiß ich gar nicht.

**Comment by Bernd9000** [comment]
Schaue gerade mal bei deiner  jarvis-info  :-)

**Comment by mcuiobroker** [comment]
Funktioniert nicht
```
{
  "20": "TOR",
  "21": "Fussball",
  "{0_userdata.0.DP.condition}=true && {val}=1": "Rainbow",
  "{0_userdata.0.DP.condition}=false && {val}=1": "Rainbow2"
}
```
![image](https://github.com/Zefau/ioBroker.jarvis/assets/84818934/29a9ae4d-9ea1-46c9-bacf-e10803bbde78)

**Comment by Bernd9000** [comment]
Vermutlich ist das in der Anzeigevariante nicht vorgesehen.
Das könnte dann nur Zefau beantworten.

**Comment by mcuiobroker** [comment]
Beispiele für die Änderung von Devices gibt es jede Menge in der Doku. 
Da kann man sich aber sehr schnell die ganze Config zerstören. 
Ich sollte es irgendwann mal angehen.

Im Standard kann nur Zefau es anpassen. Allerdings habe ich schon mehrfach das Thema Anzeigevariante und Statesübernahme angesprochen. Bislang war ich nicht erfolgreich.

**Comment by mcuiobroker** [comment]
Zeig mal bitte die Waschmaschinen DPs und wie hast du sie in Geräte integriert.

**Comment by Bernd9000** [comment]
Okay, Danke Dir erstmal,
Bin noch beim Aufbau, Maschine ist neu.
Die DPs füge ich manuell hinzu, wie fast immer.
Muß auch noch den Home Connect Adapter verstehen......gar nicht so einfach

**Comment by mcuiobroker** [comment]
Hast du alle DPs in einem Gerät in Jarvis zugeordnet? Werden die DPs in Jarvis mehrfach in unterschiedlichen Geräten genutzt?

**Comment by mcuiobroker** [comment]
Schau mal hier:
https://mcuiobroker.gitbook.io/jarvis-infos/jarvis-v3/system-effektprog./changedstatesupdatejarvis

**Comment by Bernd9000** [comment]
Super, Danke, bin gerade am testen.
Hier gehts weiter:
[IOB](https://forum.iobroker.net/topic/52228/howto-zusatz-programme-fuer-jarvis-v3/503?page=26)

**Reply by mcuiobroker** [reply]
Läuft?

**Comment by Bernd9000** [comment]
Kann ich noch nicht sagen.
